File: tests/layout_xml_mods_test.c
/*
 * OpenCReports test
 * Copyright (C) 2019-2025 Zoltán Böszörményi <zboszor@gmail.com>
 * See COPYING.LGPLv3 in the toplevel directory.
 */

#include <stdio.h>
#include <stdlib.h>

#include <opencreport.h>
#include "test_common.h"

const char *array[1][5] = {
	{ "id", "name", "property", "age", "adult" },
};

const int32_t coltypes[5] = {
	OCRPT_RESULT_NUMBER, OCRPT_RESULT_STRING, OCRPT_RESULT_STRING, OCRPT_RESULT_NUMBER, OCRPT_RESULT_NUMBER
};

int main(int argc, char **argv) {
	opencreport *o = ocrpt_init();

	if (!ocrpt_parse_xml(o, "layout_xml_test.xml")) {
		printf("XML parse error\n");
		ocrpt_free(o);
		return 0;
	}

	/*
	 * Modify all horizontal lines in <NoData> to use a custom color.
	 */
	ocrpt_report *r = get_first_report(o);
	ocrpt_output *output = ocrpt_layout_report_nodata(r);
	void *iter = NULL;
	ocrpt_output_element *elem;

	while ((elem = ocrpt_output_iterate_elements(output, &iter))) {
		if (ocrpt_output_element_is_hline(elem)) {
			ocrpt_hline *hline = (ocrpt_hline *)elem;

			ocrpt_hline_set_color(hline, "'green'");
		} else if (ocrpt_output_element_is_line(elem)) {
			/* Modify all text entries in the line to a custom string. */
			ocrpt_line *line = (ocrpt_line *)elem;
			void *iter1 = NULL;
			ocrpt_line_element *lelem;

			while ((lelem = ocrpt_line_iterate_elements(line, &iter1))) {
				if (ocrpt_line_element_is_text(lelem)) {
					ocrpt_text *text = (ocrpt_text *)lelem;

					ocrpt_text_set_value_string(text, "xxx");
				}
			}
		}
	}

	ocrpt_set_output_format(o, argc >= 2 ? atoi(argv[1]) : OCRPT_OUTPUT_PDF);

	ocrpt_execute(o);

	ocrpt_spool(o);

	ocrpt_free(o);

	return 0;
}

File: tests/layout_xml_mods_test.php
<?php
/*
 * OpenCReports test
 * Copyright (C) 2019-2025 Zoltán Böszörményi <zboszor@gmail.com>
 * See COPYING.LGPLv3 in the toplevel directory.
 */

require_once 'test_common.php';

$array = [
	[ "id", "name", "property", "age", "adult" ]
];

$coltypes = [
	OpenCReport::RESULT_NUMBER, OpenCReport::RESULT_STRING, OpenCReport::RESULT_STRING, OpenCReport::RESULT_NUMBER, OpenCReport::RESULT_NUMBER
];

$o = new OpenCReport();

if (!$o->parse_xml("layout_xml_test.xml")) {
	echo "XML parse error" . PHP_EOL;
	exit(0);
}

$r = get_first_report($o);
$out = $r->nodata();

for ($e = $out->get_first_element(); $e instanceof OpenCReport\OutputElement; $e = $e->get_next()) {
	if ($e->is_hline()) {
		$hline = $e->get_hline();
		$hline->set_color("'green'");
	} else if ($e->is_line()) {
		$l = $e->get_line();
		for ($le = $l->get_first_element(); $le instanceof OpenCReport\LineElement; $le = $le->get_next()) {
			if ($le->is_text()) {
				$t = $le->get_text();
				$t->set_value_string("xxx");
			}
		}
	}
}

$o->set_output_format(OpenCReport::OUTPUT_PDF);
$o->execute();
$o->spool();
